src/matrix/room/timeline/Timeline.js

```diff
@@ -236,8 +247,50 @@ export class Timeline {
     /** @package */
     addEntries(newEntries) {
         this._addLocalRelationsToNewRemoteEntries(newEntries);
+        this._updateFetchedEntries(newEntries);
+        this._loadRelatedEvents(newEntries);
         this._remoteEntries.setManySorted(newEntries);
     }
+
+    _updateFetchedEntries(entries) {
+        for (const entry of entries) {
+            const relatedEntry = this._fetchedEventEntries.find(e => e.id === entry.relatedEventId);
+            if (relatedEntry?.addLocalRelation(entry)) {
+                relatedEntry.dependents.forEach(e => this._findAndUpdateRelatedEntry(null, e.id, () => true));
+            }
+        }
+    }
+
+    async _loadRelatedEvents(entries) {
+        const filteredEntries = entries.filter(e => !!e.contextEventId);
+        for (const entry of filteredEntries) {
+            const id = entry.contextEventId;
+            let needToTrack = false;
+            let contextEvent;
+            // find in remote events
+            contextEvent = this.getByEventId(id);
+            // find in storage
+            if (!contextEvent) {
+                needToTrack = true;
+                contextEvent = await this._fetchEventFromStorage(id);
+            }
+            // fetch from hs
+            if (!contextEvent) {
+                contextEvent = await this._fetchEventFromHomeserver(id);
+            }
+            if (contextEvent) {
+                if (needToTrack) {
+                    // this entry was created from storage/hs, so it's not tracked by remoteEntries
+                    // we track them here so that we can update reply preview of dependents on redaction
+                    this._fetchedEventEntries.push(contextEvent);
+                }
+                contextEvent.addDependent(entry);
+                entry.setContextEntry(contextEvent);
+                // emit this change
+                this._findAndUpdateRelatedEntry(null, entry.id, () => true);
```

Comment:
```suggestion
                const txnId = entry.isPending? entry.id : null;
                const eventId = entry.isPending? null : entry.id;
                this._findAndUpdateRelatedEntry(txnId, eventId, () => true);
```

Reply:
Can we rename _findAndUpdateRelatedEntry to _findAndUpdateEntryById (and the local variables called related... therein)? It doesn't really do anything specific for relations and the name is a bit confusing here.

Reply:
Also, you could put the above in a `_updateEntry` method.